### DISCUSSION TITLE
Borderless windows connecting to xrdp on raspberry

### DISCUSSION THREAD

**Discussion by coppersoft** [neutrinolabs/xrdp#2050] [Q&A]
Hello all,

I installed xrdb on a raspbian on Raspberry pi 400. When I connect from remmina (from Linux Mint) or microsoft remote desktop (from MacOS Big Sur) all the windows are borderless and without close/resize buttons.

Please see the attached picture.

<img width="799" alt="Schermata 2021-11-10 alle 09 07 52" src="https://user-images.githubusercontent.com/7433095/141136987-f8487a24-4f77-444c-8a8c-70f9f5cdf072.png">

Do you have any idea to solve the issue?

Thank you in advance!

Lorenzo

**Comment by matt335672** [comment]
This looks like a duplicate of #2048

What is the output from `mutter --replace`?

**Comment by matt335672** [comment]
@roygoraposonjr and I have been investigating this under #2048. Here's a summary of the situation, and recommended workarounds.

# Technical description
Raspberry PI OS based on Bullseye is running xorgxrdp v0.2.12. This version of xorgxrdp has an [issue](/neutrinolabs/xorgxrdp/issues/156) where it will report an error if it is asked to resize the screen to the same size it is already configured at.

This logic is used by mutter version 3.38.6-2~deb11u1+rpt1, which is the version of mutter in Raspberry PI OS. Consequently, the versions of mutter and xorgxrdp delivered for Raspberry PI OS are incompatible.

## Affected configurations
You will run into this problem if:-
- You are running Raspberry PI OS based on Bullseye
- You have 2GB or RAM or more
- You are trying to run xrdp with the default Xorg back end.

# Possible workarounds
## Simple
- Install `tigervnc-standalone-server` and use the Xvnc back end to xrdp.

## Complex
- Build and install xorgxrdp v0.2.14 (or later). This has the [fix](/neutrinolabs/xorgxrdp/issues/161) for the resize problem. Only consider this option if you are comfortable with building software. If you go down this route it is highly recommended you also upgrade xrdp.

## Intermediate
- Configure the PI to fall back to the openbox window manager for xrdp connections.

The rest of this post details how to configure the PI to use openbox rather than mutter for xrdp sessions.

# Configure the PI to use openbox for xrdp connections
## Introduction
The raspberry PI desktop session is started by the script `/usr/bin/startlxde-pi`. As well as containing some initialisation code, this script decides whether to start the desktop with the mutter window manager or the openbox window manager.

One of the factors used to decide which window manager to use is the amount of memory installed in the PI. If less than 2GB is installed, openbox is always used.

The amount of memory in the PI is determined with this PI-specific command:-

```
vcgencmd get_config total_mem 
```

## Create another script
On your PI, create a file `/usr/bin/startlxde-pi-xrdp` with the following contents:-

```sh
#!/bin/sh
# See https://github.com/neutrinolabs/xrdp/discussions/2050

if [ -n "$XRDP_SESSION" ]; then
    # Override vcgencmd to return no memory. This causes the
    # openbox window manager to be used instead of mutter
    vcgencmd()
    {
        if [ $1/$2 = get_config/total_mem ]; then
            echo total_mem=0
        else
            echo "** Fix stub vcgencmd for $*" >&2
            false
        fi
    }
fi

. /usr/bin/startlxde-pi
```

Make the script executable with `sudo chmod +x /usr/bin/startlxde-pi-xrdp`.

What does this script do? Simply put, it checks to see if we are starting an xrdp session by checking the XRDP_SESSION environment variable. If this is set, it defines a function which overrides the `vcgencmd get_config_total_mem` command.

The script then reads the normal script to start the desktop with the shell `.` command. Because we read the script in this way, our override is in place when the normal script tries to work out how much memory is on the machine.

The net effect is that this script works exactly the same as `/usr/bin/startlxde-pi` for non-xrdp sessions, but for xrdp sessions the openbox window manager is always used.

## Set up the script for a single user
If you've only got one user on your PI, you can configure the PI to use this script for X sessions with this comand:-

```
ln -sf /usr/bin/startlxde-pi-xrdp ~/.xsession
```

If you've got more than one user you can set this script up as the default way to start the X session as follows:-

<pre>
$ <b>sudo update-alternatives --install /usr/bin/x-session-manager x-session-manager /usr/bin/startlxde-pi-xrdp 90</b>
$ <b>sudo update-alternatives --config x-session-manager</b>
There are 5 choices for the alternative x-session-manager (providing /usr/bin/x-session-manager).

  Selection    Path                        Priority   Status
------------------------------------------------------------
* 0            /usr/bin/startlxde-pi        90        auto mode
  1            /usr/bin/lxsession           49        manual mode
  2            /usr/bin/openbox-session     40        manual mode
  3            /usr/bin/startlxde           50        manual mode
  4            /usr/bin/startlxde-pi        90        manual mode
  5            /usr/bin/startlxde-pi-xrdp   90        manual mode

Press <enter> to keep the current choice[*], or type selection number: <b>5</b>
update-alternatives: using /usr/bin/startlxde-pi-xrdp to provide /usr/bin/x-session-manager (x-session-manager) in manual mode
</pre>

You can reverse this selection as follows:-

<pre>
$ <b>sudo update-alternatives --remove x-session-manager /usr/bin/startlxde-pi-xrdp</b>
update-alternatives: removing manually selected alternative - switching x-session-manager to auto mode
update-alternatives: using /usr/bin/startlxde-pi to provide /usr/bin/x-session-manager (x-session-manager) in auto mode
</pre>

**Reply by 6by9** [reply]
Typo? 
> Raspberry PI OS based on Bullseye is running xorgxrdp v0.9.12.

> Build and install xorgxrdp v0.2.14 (or later).

**Reply by matt335672** [reply]
Quite right - thanks. 0.9.12 replaced with 0.2.12.

**Reply by 6by9** [reply]
@XECDesign I don't know how involved picking up just that one patch is (or whether it would work), but is it worth a shot?

**Reply by matt335672** [reply]
It would be great if you chaps can pick this up - it'll take months going through Debian, and mutter isn't used by default on their LXDE desktop.

Commit d3cefacf2b2b13e894432437897397836bd5c58a applies cleanly to 0.2.12 on my test system.

**Reply by 6by9** [reply]
If it applies cleanly and fixes the issue, then it shouldn't be a big issue. However all packaging things are handled by spl237 and XECDesign, so I'll leave it up to them.

**Reply by tomshaffner** [reply]
@matt335672, your script above for a single user has an extra \` character in it, causing it to fail if copied and run. Think maybe your markdown had two \` characters at the end the line accidentally?

**Reply by matt335672** [reply]
Good spot - fixed!

**Reply by holly7787** [reply]
didn't work for me, I'm getting this error
![Errore RDP](https://user-images.githubusercontent.com/42677239/155685677-bf7f14db-51d4-4ab1-bf10-4a9ae5ba62bc.png)

**Reply by matt335672** [reply]
@holly7787 - read [this comment](https://github.com/neutrinolabs/xrdp/discussions/2050#discussioncomment-2102717) below. Could be #2060.

**Comment by XECDesign** [comment]
@spl237 is looking after all things xorg right now.

I think trying to patch xorgxrdp so that it's compatible with our version of mutter certainly makes sense.

**Comment by spl237** [comment]
I've just built backports of the bookworm versions of xrdp and xorgxrdp; they've been uploaded to the repo and should show up shortly. Hopefully this will fix this issue. (Taking the bookworm release with the patch in it seemed less likely to lead to problems than taking individual patches.)

**Reply by coppersoft** [reply]
I got an update on xrdp on my raspberry and now the border are visible! I don't know if it can be related, but after this update everything is incredibly slower than before!

**Reply by spl237** [reply]
I suspect that is due to an xrdp compatibility issue with mutter - mutter is very slow with VNC. It may be wiser to fall back to using openbox with xrdp.

**Reply by matt335672** [reply]
Mutter is working for me with these versions:-

```
$ dpkg -l \*xrdp
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version               Architecture Description
+++-==============-=====================-============-===============================================
ii  xorgxrdp       1:0.2.17-1~bpo11+rpt1 amd64        Remote Desktop Protocol (RDP) modules for X.org
ii  xrdp           0.9.17-2~bpo11+rpt1   amd64        Remote Desktop Protocol (RDP) server
```

@coppersoft - The visual effects may well be slower with Mutter, particularly over a LAN, as there is a lot of redrawing involved. There may be some way to disable the visual effects, although I can't find anything. However, if you do this, it will affect the console too.

**Comment by pumpkin-nbc** [comment]
My raspberry pie 4B installs the following version of xrdp
xorg-docs-core/stable,stable,now 1:1.7.1-1.2 all [已安装，自动]
xorg/stable,now 1:7.7+22 arm64 [已安装，自动]
xorgxrdp/stable,now 1:0.2.17-1~bpo11+rpt1 arm64 [已安装，自动]
xrdp/stable,now 0.9.17-2~bpo11+rpt1 arm64 [已安装]
xscreensaver-data/stable,now 5.45+dfsg1-2 arm64 [已安装，自动]
xscreensaver/stable,now 5.45+dfsg1-2 arm64 [已安装]
xsel/stable,now 1.2.0+git9bfc13d.20180109-3 arm64 [已安装，自动]

Why does it also show connection failure?

**Reply by matt335672** [reply]
@pumpkin-nbc - this discussion is nothing to do with connection failure. I suspect you've got an entirely different problem.

**Reply by pumpkin-nbc** [reply]
> 

Yes, I'm sorry. I didn't pay attention to this problem.
But now I have a problem with xrdp connecting raspberry pi 4B.

**Reply by matt335672** [reply]
That's fine. Please raise a separate issue though, so anyone reading this thread doesn't get too confused.

**Reply by pumpkin-nbc** [reply]
> 
I have established the problem
https://github.com/neutrinolabs/xrdp/issues/2060

**Comment by waterwin** [comment]
Even with this setup nothing has changed on my up to date Bullseye Raspberry Pi 4 B. RDC hanging in nice bluegreen screen with nice border after login
 
```
dpkg -l \*xrdp
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version               Architecture Description
+++-==============-=====================-============-===============================================
ii  xorgxrdp       1:0.2.17-1~bpo11+rpt1 armhf        Remote Desktop Protocol (RDP) modules for X.org
ii  xrdp           0.9.17-2~bpo11+rpt1   armhf        Remote Desktop Protocol (RDP) server
```

**Comment by matt335672** [comment]
@waterwin - that doesn't sound the same at all. This is a discussion about borderless Windows. Yours sounds to be more about starting a session.

There's a lot of discussion about rPI issues in  #2129, #2089 and #2060. In summary; trying to use the user `pi` over RDP is really difficult to get working because of two problems:-
1) `systemd --user` won't allow the same user on the console and RDP at the same time.
2) The kernel acceleration drivers on the PI kernel don't work with remote sessions.

There are complex ways around both of these, but the easiest solution is to set up a different user for RDP use.

**Reply by holly7787** [reply]
This works, thanks!

**Comment by waterwin** [comment]
Sorry if I misunderstand the issues.
Is your answer referring to Buster or Bullseye ?
I did not think it difficult to use RDC via XRDP from a Raspberry Pi running Buster.It runs on my 2B, 3B+ and 4BI also have a Raspberry Pi3B+ running Bullseye using XRDP
I tried to read through all issues and finally arrived via a pointer by you on #2050But the discussions sometimes start to combine different things like I did now again obviously.

Which leaves me with a Raspberry 4B Bullseye not running XRDP out of the box. I will try to get it working with another user not being pi.
Thanks for the correction.


    On Thursday, February 3, 2022, 11:23:57 AM GMT+1, matt335672 ***@***.***> wrote:  
 
 


@waterwin - that doesn't sound the same at all. This is a discussion about borderless Windows. Yours sounds to be more about starting a session.

There's a lot of discussion about rPI issues in #2129, #2089 and #2060. In summary; trying to use the user pi over RDP is really difficult to get working because of two problems:-
   
   - systemd --user won't allow the same user on the console and RDP at the same time.
   - The kernel acceleration drivers on the PI kernel don't work with remote sessions.

There are complex ways around both of these, but the easiest solution is to set up a different user for RDP use.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Reply by matt335672** [reply]
I'm talking about Bullseye. This seems to be more difficult to get working than Buster. Since you have Bullseye working on your Pi3B+, I think this is likely to be related to the kernel acceleration driver (https://github.com/neutrinolabs/xrdp/issues/2060)

**Reply by AndyF1** [reply]
None of the solutions above worked for me. What did was creating a new user, and using that user for RDP.

**Reply by mike9755** [reply]
> I'm talking about Bullseye. This seems to be more difficult to get working than Buster. Since you have Bullseye working on your Pi3B+, I think this is likely to be related to the kernel acceleration driver (#2060)

@matt335672 
Hi, I have shared some comment, @you for your attention here, I think the problem was not from Bullseye, same system can work correctly on Raspberry Pi 3B.

 I met the complete same problem with below picture, the picuture was from trying to connect via Windows remote connection by Xorg, I have looked this problem several days, seems the issue happened very long time ago and have no perfect solution till now.
I think all guys met this issue should come from hardware Raspberry Pi 4B model, my SD card was installed Bullseye 64 bit version, and works perfect with XRDP under Pi 3B model, when I move the same SD card to Pi 4B hardware then met this problem, I don't know the reason, but I think the problem only was met at Pi 4B ( note: Pi 4B has two Micro HDMI, maybe is the problem?) please check and hope someone can fix this issue as soon, thank you.

![image](https://user-images.githubusercontent.com/5673534/188665542-40467c8a-3652-49a6-b965-13035160eb76.png)

**Reply by matt335672** [reply]
@mike9755 - I can only refer you to my comment [above](https://github.com/neutrinolabs/xrdp/discussions/2050#discussioncomment-2102717).
